### ISSUE TITLE
async middleware

### ISSUE THREAD

**Issue by sophister** [yargs/yargs#1237]
hi, it seems `yargs` does not support `async` middleware? I've add an async middleware, but not work:

```typescript
import * as yargs from 'yargs';

export const command = `test`;
export const aliases = [ 'test', 't'];
export const desc = `tinker demo`;
export async function handler(argv: IArgs){
    console.log(`begin tinker: `, argv._test);
    const out = await new Promise((resolve, reject) => {
        console.log(`in promise`);
        setTimeout(() => {
            console.log('before resolve');
            resolve('hi');
        }, 1000);
    });
    console.log('after promise, reulst: ' + out);
}
export const middlewares = [ test ];

interface IArgs extends yargs.Arguments{
    _test: string;
}

async function test(argv: IArgs){
    console.log('in middleware test');
    const out = await new Promise<string>( (resolve, reject) => {
        setTimeout( () => {
            console.log('async middleware before resolve');
            resolve('async middleware');
        }, 2000);
    });
    argv._test = out;
    return argv;
}
```

**Comment by aorinevo** [comment]
HI @sophister,  I would love to see middleware work for async functions but at the moment we only support sync.  If you have the time and would like to work on a PR for this, I'd be happy to review it.

**Comment by amurdock** [comment]
Hi @aorinevo, I'm looking at putting a PR together for this but thought I'd get some feedback on the approach first. My preferred approach would be to chain middleware promises before the command is run. Since the command itself supports promises this should work reasonably seamlessly. Something similar to ...
```
const middlewares = commandHandler.middlewares || []

middlewares
  .reduce((chain, middleware) => {
    // return chained promise from middleware
  }, Promise.resolve(innerArgv))
  .then(innerArgv => {
    return commandHandler.handler(innerArgv)
  })
  .catch(error => {
    yargs.getUsageInstance().fail(null, error)
  })
```

Thoughts?

**Comment by bcoe** [comment]
@amurdock @sophister please take this for a spin:

`npm i yargs@next`

It should have support for async middleware; let me know how it's working for you, as I'd like to promote the release relatively soon.

**Comment by akervern** [comment]
@bcoe Perfect timing. I was looking for this issue as well, it works well for me!

**Comment by bcoe** [comment]
@akervern @amurdock @sophister this is now available in `yargs@13.1.0`.
